### PR TITLE
Adds services to save and load sessions

### PR DESCRIPTION
**Pull request by javamind**

### CODE REVIEW

File: src/main/kotlin/mixit/controller/SessionController.kt [hunk 1]
```diff
@@ -0,0 +1,26 @@
...
+import org.springframework.web.reactive.function.BodyInsertersExtension.fromPublisher
+import org.springframework.web.reactive.function.server.*
+import org.springframework.web.reactive.function.server.RequestPredicates.GET
+import org.springframework.web.reactive.function.server.ServerResponse.ok
+import reactor.core.publisher.Mono
+
+class SessionController(val repository: SessionRepository) : RouterFunction<ServerResponse> {
+
+    // TODO Relax generics check to avoid explicit casting
+    override fun route(request: ServerRequest) = RouterFunctions
```
Comment (sdeleuze) [inline src/main/kotlin/mixit/controller/SessionController.kt]: Could you add `@Suppress("UNCHECKED_CAST")` like for other controllers?

File: src/main/kotlin/mixit/controller/SessionController.kt [hunk 2]
```diff
@@ -0,0 +1,26 @@
...
+            .route(request) as Mono<HandlerFunction<ServerResponse>>
+
+    fun findOne() = HandlerFunction { req ->
+        ok().contentType(APPLICATION_JSON_UTF8).body(fromPublisher(repository.findOne(req.pathVariable("id"))))
+    }
+
+    fun findAll() = HandlerFunction {
+        ok().contentType(APPLICATION_JSON_UTF8).body(fromPublisher(repository.findAll()))
+    }
+}
```
Comment (sdeleuze) [inline src/main/kotlin/mixit/controller/SessionController.kt]: Let's add a new line there

File: src/main/kotlin/mixit/model/SessionLanguage.kt [hunk 3]
```diff
@@ -0,0 +1,6 @@
+package mixit.model
+
+enum class SessionLanguage {
+    fr,
+    en
+}
```
Comment (sdeleuze) [inline src/main/kotlin/mixit/model/SessionLanguage.kt]: Missing new line

File: src/main/kotlin/mixit/model/SessionFormat.kt [hunk 4]
```diff
@@ -0,0 +1,9 @@
+package mixit.model
+
+enum class SessionFormat(val duration: Int) {
+    Talk(50),
+    LightningTalk(5),
+    Workshop(110),
+    Random(25),
+    Keynote(25)
+}
```
Comment (sdeleuze) [inline src/main/kotlin/mixit/model/SessionFormat.kt]: Missing new line

File: src/main/kotlin/mixit/repository/SessionRepository.kt [hunk 5]
```diff
@@ -0,0 +1,40 @@
...
+
+        val objectMapper: ObjectMapper = Jackson2ObjectMapperBuilder.json().build()
+        var sessions: List<SessionDataDto> = objectMapper.readValue(file.file)
+
+        sessions
+                .map { session -> session.toSession() }
+                .forEach { session -> save(session).block() }
+    }
+
+}
```
Comment (sdeleuze) [inline src/main/kotlin/mixit/repository/SessionRepository.kt]: Missing new line

File: src/main/kotlin/mixit/repository/SpeakerRepository.kt [hunk 6]
```diff
@@ -0,0 +1,54 @@
...
+import java.time.LocalDate
+
+class SpeakerRepository(db: ReactiveMongoTemplate, f: ReactiveMongoRepositoryFactory, val resourceLoader: ResourceLoader) :
+        SimpleReactiveMongoRepository<Speaker, String>(f.getEntityInformation(Speaker::class), db) {
+
+
+    fun initData() {
+        deleteAll().block()
+//
+//        val events = listOf(
```
Comment (sdeleuze) [inline src/main/kotlin/mixit/repository/SpeakerRepository.kt]: Could we delete these comments? Or are they here for a specific reason?

File: src/main/kotlin/mixit/repository/SpeakerRepository.kt [hunk 7]
```diff
@@ -0,0 +1,54 @@
...
+//        val objectMapper: ObjectMapper = Jackson2ObjectMapperBuilder.json().build()
+//        var sponsors: List<MemberDataDto> = objectMapper.readValue(file.file)
+//
+////        val ev = event.copy();
+////        ev.sponsors = sponsors.flatMap { sponsor -> sponsor.toEventSponsoring(event.year) }
+////        return ev;
+//        return Speaker()
+//    }
+
+}
```
Comment (sdeleuze) [inline src/main/kotlin/mixit/repository/SpeakerRepository.kt]: Missing new line

File: src/main/resources/data/speaker/speaker_2012.json [hunk 8]
```diff
@@ -0,0 +1 @@
...pedia depuis 2009","level":[{"key":"","value":null}]},{"idMember":274,"login":"bodil","firstname":"Bodil","lastname":"STOKKE","logo":null,"hash":"f9f8e0403b0437662bdbc2b4e9c6ccf3","shortDescription":"Internet Scientist","level":[{"key":"","value":null}]},{"idMember":300,"login":"jrf","firstname":"Julien","lastname":"RICHARD-FOY","logo":null,"hash":"c4c26da9c36518e9841ae30f276f64e6","shortDescription":"Développeur","level":[{"key":"","value":null}]},{"idMember":67,"login":"rauschma","firstname":"Axel","lastname":"RAUSCHMAYER","logo":null,"hash":"5592487d2231fbe06ed2ac80d92c8f20","shortDescription":"Dr. Axel Rauschmayer is a consultant and trainer for JavaScript, web technologies and information management.","level":[{"key":"","value":null}]},{"idMember":84,"login":"alain.delafosse","firstname":"Alain","lastname":"DELAFOSSE","logo":null,"hash":"bc91e6b6e5565e352977abfcb20277c2","shortDescription":"Chief Architect at Kelkoo","level":[{"key":"","value":null}]},{"idMember":379,"login":"capponi.nicolas","firstname":"Nicolas","lastname":"CAPPONI","logo":null,"hash":"d566bb46fa9bb296a8cec9e9d3bac644","shortDescription":"Développeur / Architecte à Kelkoo","level":[{"key":"","value":null}]}]
```
Comment (sdeleuze) [inline src/main/resources/data/speaker/speaker_2012.json]: Missing new line

File: src/main/resources/data/speaker/speaker_2013.json [hunk 9]
```diff
@@ -0,0 +1 @@
...,"lastname":"KALDERIMIS","logo":null,"hash":"f45c05ad2fca21e73fe4010a35b61f87","shortDescription":"Shipping technical debt daily!\nOh, and I work for Travis.","level":[{"key":"","value":null}]},{"idMember":776,"login":"kurichan","firstname":"André","lastname":"ESPAZE","logo":null,"hash":"b784756b9263f76d75326ed37315ad48","shortDescription":"membre du /tmp/lab, hackerspace créé en 2007 et d'Usinette.org, hackerspace spécifiquement dédié à la construction numérique.","level":[{"key":"","value":null}]},{"idMember":829,"login":"martinez@lyon.cci.fr","firstname":"Christian","lastname":"MARTINEZ","logo":null,"hash":"694dc8805b1a1e967af755ce85d6cd81","shortDescription":"Concepteur et développeur de jeux formatifs","level":[{"key":"","value":null}]},{"idMember":49,"login":"bootis","firstname":"Aline","lastname":"PAPONAUD","logo":null,"hash":"d5d0721435ade58bb9ae5dab6eb9b091","shortDescription":"Développeuse chez Sfeir","level":[{"key":"","value":null}]},{"idMember":372,"login":"neveu.audrey","firstname":"Audrey","lastname":"NEVEU","logo":null,"hash":"7e357ae5b0366a29f1f54434791ddefd","shortDescription":"Jug Leader JDuchess et développeur Java chez Sfeir","level":[{"key":"","value":null}]}]
```
Comment (sdeleuze) [inline src/main/resources/data/speaker/speaker_2013.json]: Missing new line

File: src/main/resources/data/staff.json [hunk 10]
```diff
@@ -0,0 +1 @@
...iption":"Indépendant. Dev & Ops, Formation.","level":[]},{"idMember":68,"login":"mpetitdant","firstname":"Mathieu","lastname":"PETITDANT","logo":null,"hash":"5865a6e299822c23b4bcc9204d4e22d4","shortDescription":"Développeur Java/Web et agiliste ... freelance.","level":[]},{"idMember":252,"login":"guillaumeehret","firstname":"Guillaume","lastname":"EHRET","logo":null,"hash":"aa4d47d1016e45c23b6af05ec11c0a9c","shortDescription":"Développeur agile passionné par la technique et l'agilité, fondateur de [Dev-Mind](https://dev-mind.fr/)","level":[{"key":"","value":null}]},{"idMember":666,"login":"sdeleuze","firstname":"Sébastien","lastname":"DELEUZE","logo":null,"hash":"04bf197ddd198f4ff3a50e93ebd37272","shortDescription":"Membre de l'équipe de développement de Spring Framework et de l'équipe d'orga de Mix-IT.","level":[{"key":"","value":null}]},{"idMember":5142,"login":"92989140","firstname":"Grégory","lastname":"ALEXANDRE","logo":null,"hash":"91c77f73775c974ef787492a2193d2c6","shortDescription":"Agiliste depuis plus de 8 ans et coach Agile chez [Goood!](http://goood.pro), GO de la communauté agile lyonnaise via Mix-IT et le [CARA Lyon](http://lyon.clubagilerhonealpes.org/)","level":[]}]
```
Comment (sdeleuze) [inline src/main/resources/data/staff.json]: Missing new line

File: src/test/kotlin/mixit/integration/SessionEndpointSpec.kt [hunk 11]
```diff
@@ -0,0 +1,42 @@
...
+
+class SessionEndpointSpec : SubjectSpek<Application>({
+
+    val webClient = WebClient.builder(ReactorClientHttpConnector()).build()
+
+    subject {
+        Application(port = SocketUtils.findAvailableTcpPort())
+    }
+
+    //TODO
```
Comment (sdeleuze) [inline src/test/kotlin/mixit/integration/SessionEndpointSpec.kt]: Would be nice to have at least a test ;-)

**Comment by javamind** [comment]
@sdeleuze can you review this code ?

File: src/main/kotlin/mixit/repository/EventRepository.kt [hunk 12]
```diff
@@ -40,7 +40,7 @@ class EventRepository(db: ReactiveMongoTemplate, f: ReactiveMongoRepositoryFacto
      * Loads data from the json sponsor files
      */
     fun readSponsorsForEvent(event: Event): Event {
-        val filename = "classpath:data/sponsor/sponsor_".plus(event.year).plus(".json")
+        val filename = "classpath:data/sponsor/sponsor_mixit".plus(event.id).plus(".json")
```
Comment (sdeleuze) [inline src/main/kotlin/mixit/repository/EventRepository.kt]: Use string interpolation instead

File: src/main/kotlin/mixit/data/dto/SessionDataDto.kt [hunk 13]
```diff
@@ -30,7 +30,7 @@ data class SessionDataDto(
                 Room.findByName(room ?: ""),
                 start,
                 end,
-                "sess" + idSession
+                "".plus(idSession)
```
Comment (sdeleuze) [inline src/main/kotlin/mixit/data/dto/SessionDataDto.kt]: Use string interpolation instead

File: src/main/kotlin/mixit/repository/SessionRepository.kt [hunk 14]
```diff
@@ -7,26 +7,30 @@ import mixit.model.Session
...
     }
 
     /**
-     * Loads data from the json sponsor files
+     * Loads data from the json session files
      */
-    fun saveSessionForEvent(year: Int) {
-        val filename = "classpath:data/session/session_".plus(year).plus(".json")
+    private fun saveSessionsByYear(year: Int) {
+        val filename = "classpath:data/session/session_mixit".plus(year).plus(".json")
```
Comment (sdeleuze) [inline src/main/kotlin/mixit/repository/SessionRepository.kt]: Use string interpolation instead

File: src/main/kotlin/mixit/repository/SpeakerRepository.kt [hunk 15]
```diff
@@ -3,52 +3,45 @@ package mixit.repository
...
-////        val ev = event.copy();
-////        ev.sponsors = sponsors.flatMap { sponsor -> sponsor.toEventSponsoring(event.year) }
-////        return ev;
-//        return Speaker()
-//    }
-
-}
\ No newline at end of file
+    private fun saveSpeakersByYear(year: Int) {
+        val filename = "classpath:data/speaker/speaker_mixit".plus(year).plus(".json")
```
Comment (sdeleuze) [inline src/main/kotlin/mixit/repository/SpeakerRepository.kt]: Use string interpolation instead

File: src/main/kotlin/mixit/support/MixitApi.kt [hunk 16]
```diff
@@ -0,0 +1,12 @@
+package mixit.support
+
+class MixitApi {
+    companion object {
+        /**
+         * The event ids are the year on 2 digits. This function converts an id to a year on 4 digits
+         */
+        fun idToYear(id: String): Int {
+            return "20".plus(id).toInt()
```
Comment (sdeleuze) [inline src/main/kotlin/mixit/support/MixitApi.kt]: Use string interpolation instead

File: src/main/resources/data/speaker/speaker_mixit12.json [hunk 17]
```diff
@@ -0,0 +1 @@
...ata."},{"idMember":289,"login":"kali","firstname":"Mathieu","lastname":"POUMEYROL","logo":null,"hash":"4142d70056c184c89c6544027a6ac89e","shortDescription":"Pelleteur de nuage — ou Cloud Engineer — à Fotopedia depuis 2009"},{"idMember":274,"login":"bodil","firstname":"Bodil","lastname":"STOKKE","logo":null,"hash":"f9f8e0403b0437662bdbc2b4e9c6ccf3","shortDescription":"Internet Scientist"},{"idMember":300,"login":"jrf","firstname":"Julien","lastname":"RICHARD-FOY","logo":null,"hash":"c4c26da9c36518e9841ae30f276f64e6","shortDescription":"Développeur"},{"idMember":67,"login":"rauschma","firstname":"Axel","lastname":"RAUSCHMAYER","logo":null,"hash":"5592487d2231fbe06ed2ac80d92c8f20","shortDescription":"Dr. Axel Rauschmayer is a consultant and trainer for JavaScript, web technologies and information management."},{"idMember":84,"login":"alain.delafosse","firstname":"Alain","lastname":"DELAFOSSE","logo":null,"hash":"bc91e6b6e5565e352977abfcb20277c2","shortDescription":"Chief Architect at Kelkoo"},{"idMember":379,"login":"capponi.nicolas","firstname":"Nicolas","lastname":"CAPPONI","logo":null,"hash":"d566bb46fa9bb296a8cec9e9d3bac644","shortDescription":"Développeur / Architecte à Kelkoo"}]
```
Comment (sdeleuze) [inline src/main/resources/data/speaker/speaker_mixit12.json]: Missing new line

File: src/main/resources/data/speaker/speaker_mixit13.json [hunk 18]
```diff
@@ -0,0 +1 @@
...CKI","logo":null,"hash":"307999df7c27deafc2818fdcad76c8b4","shortDescription":"node.js engineer, working for nodejitsu."},{"idMember":672,"login":"j2h","firstname":"Josh","lastname":"KALDERIMIS","logo":null,"hash":"f45c05ad2fca21e73fe4010a35b61f87","shortDescription":"Shipping technical debt daily!\nOh, and I work for Travis."},{"idMember":776,"login":"kurichan","firstname":"André","lastname":"ESPAZE","logo":null,"hash":"b784756b9263f76d75326ed37315ad48","shortDescription":"membre du /tmp/lab, hackerspace créé en 2007 et d'Usinette.org, hackerspace spécifiquement dédié à la construction numérique."},{"idMember":829,"login":"martinez@lyon.cci.fr","firstname":"Christian","lastname":"MARTINEZ","logo":null,"hash":"694dc8805b1a1e967af755ce85d6cd81","shortDescription":"Concepteur et développeur de jeux formatifs"},{"idMember":49,"login":"bootis","firstname":"Aline","lastname":"PAPONAUD","logo":null,"hash":"d5d0721435ade58bb9ae5dab6eb9b091","shortDescription":"Développeuse chez Sfeir"},{"idMember":372,"login":"neveu.audrey","firstname":"Audrey","lastname":"NEVEU","logo":null,"hash":"7e357ae5b0366a29f1f54434791ddefd","shortDescription":"Jug Leader JDuchess et développeur Java chez Sfeir"}]
```
Comment (sdeleuze) [inline src/main/resources/data/speaker/speaker_mixit13.json]: Missing new line

**Pull request merged**